### PR TITLE
Feature/object mapping

### PR DESCRIPTION
**Pull request by RainbowDashLabs**
Adds a minimal implementation for mapping rows to objects.

The object is defined by the query builder already. The mapper will be determined by the names or labels of the columns in the result set. A class can have multiple mappers for different sets of column names.

Minimal example:
```java
// Create a row mapper for the Result class with three different colums
RowMapper<Result> fullMapper = RowMapper.forClass(Result.class)
                                        // Define how the row should be mapped
                                        .mapper(row -> new MetaResult(row.getInt("id"),
                                                row.getString("result"),
                                                row.getString("meta")))
                                        // define the column names
                                        .addColumns("id", "result", "meta")
                                        .build();

// Allows polymorphism
RowMapper<Result> sparseMapper = RowMapper.forClass(Result.class)
                                          // Define how the row should be mapped
                                          .mapper(row -> new Result(row.getInt("id"),
                                                  row.getString("result")))
                                          // define the column names
                                          .addColumns("id", "result")
                                          .build();

// Register the mapper
RowMappers rowMappers = new RowMappers().register(fullMapper, sparseMapper);

// Retrieves a list of MetaResults
List<Result> metaResults = QueryBuilder.builder(source, Result.class)
        .defaultConfig(config -> config.rowMappers(rowMappers))
        .query("SELECT id, result, meta FROM results")
        .emptyParams()
        // Call map instead of read rows. This will let the query builder determine the type by itself.
        .map()
        .allSync();


// Retrieves a list of Results
List<Result> results = QueryBuilder.builder(source, Result.class)
        .defaultConfig(config -> config.rowMappers(rowMappers))
        .query("SELECT id, result FROM results")
        .emptyParams()
        // Call map instead of read rows. This will let the query builder determine the type by itself.
        .map()
        .allSync();

// Allows setting of an alias for a column without the need of creating a new mapper
results = QueryBuilder.builder(source, Result.class)
        .defaultConfig(config -> config.rowMappers(rowMappers))
        .query("SELECT id, result as r_result FROM results")
        .emptyParams()
        // Call map instead of read rows. We map the column result to r_result when it gets requested.
        .map(conf -> conf.addAlias("result", "r_result"))
        .allSync();
```

### CODE REVIEW

File: sadu-queries/src/test/java/de/chojo/sadu/wrapper/mapper/Result.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
+/*
+ *     SPDX-License-Identifier: AGPL-3.0-only
+ *
+ *     Copyright (C) 2022 RainbowDashLabs and Contributor
+ */
+
+package de.chojo.sadu.wrapper.mapper;
+
+class Result {
```
Comment (Goldmensch) [inline sadu-queries/src/test/java/de/chojo/sadu/wrapper/mapper/Result.java:9]: Just use a record here.

File: sadu-queries/src/main/java/de/chojo/sadu/wrapper/mapper/RowMappers.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Class to register {@link RowMapper} to map rows to objects.
+ */
+public class RowMappers {
```
Comment (Goldmensch) [inline sadu-queries/src/main/java/de/chojo/sadu/wrapper/mapper/RowMappers.java:27]: I think "MapperRegistry" similar would be less confusing. RowMappers sounds like a name of a field of type list.

**Review by Goldmensch — CHANGES_REQUESTED** [review]

File: sadu-queries/src/main/java/de/chojo/sadu/wrapper/QueryBuilder.java [hunk 3]
```diff
@@ -141,14 +157,21 @@ public ResultStage<T> parameter(ThrowingConsumer<ParamBuilder, SQLException> par
 
     @Override
     public RetrievalStage<T> readRow(ThrowingFunction<T, Row, SQLException> mapper) {
-        currResultMapper = mapper;
+        currRowMapper = RowMapper.forClass(clazz).mapper(mapper).build();
```
[the remark this answers is not in the document]
Comment (RainbowDashLabs) [reply sadu-queries/src/main/java/de/chojo/sadu/wrapper/QueryBuilder.java:160]: If the class is null the return type of the method is void. Also the class is not annotated as nullable and therefore using null may end up in an exception which is expected.

I will add an assertion here.

File: sadu-queries/src/main/java/de/chojo/sadu/wrapper/QueryBuilderConfig.java [hunk 4]
```diff
@@ -53,6 +59,7 @@ public static Builder builder() {
      * @param config config to set
      */
     public static void setDefault(QueryBuilderConfig config) {
```
[the remark this answers is not in the document]
Comment (RainbowDashLabs) [reply sadu-queries/src/main/java/de/chojo/sadu/wrapper/QueryBuilderConfig.java:61]: Most user will, but users like me go with the inheritance approach, which in the end will let them end up with multiple QueryBuilders in multiple classes. In this case this spares me another reconfiguration on every query builder.

The default config is also used when modifying the config via the defaultConfig(Consumer) call.

File: sadu-queries/src/main/java/de/chojo/sadu/wrapper/mapper/RowMappers.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+     * A class can only have one mapper per column combination.
+     *
+     * @param rowMapper the mapper to register
+     * @throws MappingAlreadyRegisteredException when a mapping with the same column name exists already
+     */
+    public RowMappers register(RowMapper<?> rowMapper) {
+        var rowMappers = mapper.computeIfAbsent(rowMapper.clazz(), key -> new ArrayList<>());
+        for (var mapper : rowMappers) {
+            if (mapper.columns().containsAll(rowMapper.columns())
+                && mapper.columns().size() == rowMapper.columns().size()) {
```
[the remark this answers is not in the document]
Comment (RainbowDashLabs) [reply sadu-queries/src/main/java/de/chojo/sadu/wrapper/mapper/RowMappers.java:42]: I just dont trust equals on collections in general I think, but true :D

File: sadu-queries/src/main/java/de/chojo/sadu/wrapper/mapper/rowmapper/RowMapper.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    private final ThrowingFunction<? extends T, Row, SQLException> mapper;
+    private final Set<String> columns;
+
+    RowMapper(Class<T> clazz, ThrowingFunction<? extends T, Row, SQLException> mapper, Set<String> columns) {
+        this.clazz = clazz;
+        this.mapper = mapper;
+        this.columns = columns;
+    }
+
+    public static <T> PartialRowMapper<T> forClass(Class<T> clazz) {
```
[the remark this answers is not in the document]
Comment (RainbowDashLabs) [reply sadu-queries/src/main/java/de/chojo/sadu/wrapper/mapper/rowmapper/RowMapper.java:40]: the RowMapperBuilder is package private. This method is used to force correct chaining. A partial build has no build method.

File: sadu-queries/src/test/java/de/chojo/sadu/wrapper/mapper/Result.java (hunk 1 above)
Comment (RainbowDashLabs) [reply sadu-queries/src/test/java/de/chojo/sadu/wrapper/mapper/Result.java:9]: I lowered the target language level to 15 again.

File: build.gradle.kts [hunk 7]
```diff
@@ -39,12 +39,13 @@ allprojects {
     java {
         withSourcesJar()
         withJavadocJar()
-        sourceCompatibility = JavaVersion.VERSION_15
+        sourceCompatibility = JavaVersion.VERSION_17
```
Comment (Amejonah1200) [inline build.gradle.kts]: Why no toolchain?

File: sadu-queries/src/main/java/de/chojo/sadu/wrapper/mapper/RowMappers.java (hunk 2 above)
Comment (RainbowDashLabs) [reply sadu-queries/src/main/java/de/chojo/sadu/wrapper/mapper/RowMappers.java:27]: Renamed it to RowMappersRegistry

File: sadu-queries/src/main/java/de/chojo/sadu/wrapper/QueryBuilder.java [hunk 8]
```diff
@@ -62,13 +67,16 @@
     private final Queue<QueryTask> tasks = new ArrayDeque<>();
     private final QueryExecutionException executionException;
     private final WrappedQueryExecutionException wrappedExecutionException;
+    private final Class<T> clazz;
     private String currQuery;
     private ThrowingConsumer<PreparedStatement, SQLException> currStatementConsumer;
-    private ThrowingFunction<T, Row, SQLException> currResultMapper;
+    private RowMapper<T> currRowMapper;
```
[the remark this answers is not in the document]
Comment (RainbowDashLabs) [reply sadu-queries/src/main/java/de/chojo/sadu/wrapper/QueryBuilder.java:73]: It does change with each stage. The update, insert and delete statements set it to null in the meantime. Although it is not possible to chain another query after a select it is still used.

**Review by Goldmensch — APPROVED** [review]

**Review by Goldmensch — APPROVED** [review]

**Pull request merged**